xcm/xcm-builder/src/weight.rs

```diff
@@ -72,27 +74,105 @@ impl<T: Get<Weight>, C: Decode + GetDispatchInfo> WeightBounds<C> for FixedWeigh
 	}
 }
 
+/// Function trait for handling some revenue. Similar to a negative imbalance (credit) handler, but for a
+/// `MultiAsset`. Sensible implementations will deposit the asset in some known treasury or block-author account.
+pub trait TakeRevenue {
+	/// Do something with the given `revenue`, which is a single non-wildcard `MultiAsset`.
+	fn take_revenue(revenue: MultiAsset);
+}
+
+/// Null implementation just burns the revenue.
+impl TakeRevenue for () {
+	fn take_revenue(_revenue: MultiAsset) {}
+}
+
 /// Simple fee calculator that requires payment in a single concrete fungible at a fixed rate.
 ///
 /// The constant `Get` type parameter should be the concrete fungible ID and the amount of it required for
 /// one second of weight.
-pub struct FixedRateOfConcreteFungible<T>(Weight, PhantomData<T>);
-impl<T: Get<(MultiLocation, u128)>> WeightTrader for FixedRateOfConcreteFungible<T> {
-	fn new() -> Self { Self(0, PhantomData) }
-	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, ()> {
+pub struct FixedRateOfConcreteFungible<
+	T: Get<(MultiLocation, u128)>,
+	R: TakeRevenue,
+>(Weight, u128, PhantomData<(T, R)>);
+impl<T: Get<(MultiLocation, u128)>, R: TakeRevenue> WeightTrader for FixedRateOfConcreteFungible<T, R> {
+	fn new() -> Self { Self(0, 0, PhantomData) }
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, Error> {
 		let (id, units_per_second) = T::get();
 		let amount = units_per_second * (weight as u128) / 1_000_000_000_000u128;
```

Comment:
```suggestion
		let amount = units_per_second * (weight as u128) / frame_support::weights::constants::WEIGHTS_PER_NANO;
```

or is this independent? Either way better to make it a const rather than a magic number.

Reply:
@kianenigma should be `WEIGHT_PER_SECOND`. Updated for you.